### PR TITLE
Added Main environment class and main function

### PR DESCRIPTION
**Pull request by kexinrong**

### CODE REVIEW

**Comment by adbh94** [comment]
We still need to standardize our measurement collections. How about the environment have an enum of all the metrics it expects to collect? Then, each device can report to it a dictionary with keys identifying a metric and values being the actual measurements.

**Comment by kexinrong** [comment]
@sdhpzhtk I'm in the middle of changing the Link class and i'm wondering what to do with the report function.

**Comment by sdhpzhtk** [comment]
We do not need a enum in main. Just return a dictionary. (dict key should be a descriptive string)

**Comment by adbh94** [comment]
I really think having an enum in the environment will make things much cleaner. It standardizes the report funcion and clearly marks all the quantities you expect to measure in the simulation.

**Comment by kexinrong** [comment]
@sdhpzhtk progress?

**Comment by sdhpzhtk** [comment]
@kexinrong  Oh I forgot to git push...

File: env.py [hunk 1]
```diff
@@ -1,2 +1,148 @@
 import simpy
+import time
 
+from output.py import *
+from input_network import *
+from host import *
+from link import *
+from flow import *
```
Comment (adbh94) [inline env.py]: Please avoid using import \* when you can explicitly state what exactly you want to import. http://stackoverflow.com/questions/2386714/why-is-import-bad

File: env.py [hunk 2]
```diff
@@ -1,2 +1,148 @@
...
+                    
+    FLOW_FIELDS = ['flow_send_rate',
+                   'flow_receive_rate,
+                   'flow_avg_RTT',
+                  ]
+        
+    LINK_FIELDS = ['packet_loss',
+                   'buffer_occupancy',
+                   'link_rate',
+                  ]
```
Comment (adbh94) [inline env.py:24]: You're enumerating the metrics here... except you're not using an enum... I really think we should have an enum called Measurements containing these parameters.

File: env.py [hunk 3]
```diff
@@ -1,2 +1,148 @@
...
+        Attrs:
+            hosts: a list of host objs
+            flows: a list of flow objs
+            routers: a list of router objs
+            links: a list of link objs
+            duration: user specified duration of simulation (in ms)
+            interval: interval that env collects data at (in ms)
+            realTimeGraph: realTimeGraph obj
+            maxId: the max ID the network has assgined to any objs
+        '''
```
Comment (adbh94) [inline env.py]: We've adopted an indentation scheme for the docstrings. Refer to the other files to see how things are tabbed under Args and Attributes.

File: env.py (hunk 2 above)
Comment (sdhpzhtk) [inline env.py:24]: No enum type in py2.

Also I feel like ‘host_send_rate’ looks much cleaner than env.host_send_rate or env.Measurement.host_send_rate

郑韬堃 Taokun (Xander) Zheng

Email: tzheng@caltech.edu
Mobile: 1-626-354-2788

Address:
California Institute of Technology
MSC 1076, Pasadena, CA 91126

**Comment by aman-agarwal** [comment]
There some pending issues here.

File: env.py (hunk 2 above)
Comment (adbh94) [inline env.py:24]: Enums are just static classes, so they can be easily implemented in py2. It's done in host.py for example. Anyways, I've modified the host code to not rely on a Measurements enum, so we'll go with your way for now.

**Comment by sdhpzhtk** [comment]
@kexinrong @adbh94 @aman-agarwal Code review? Note the command line main function is in simulator.py

File: simulator.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+'''
+The main function for the network simulator for command line
+'''
+
+import sys, getopt
+from env import *
```
Comment (adbh94) [inline simulator.py]: Here as well, specify what exactly you are importing.

File: simulator.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		print 'Total duration should be a positive int'
+		sys.exit(2)
+	if interval <= 0:
+		print 'Interval for data collection should be a positive int'
+		sys.exit(2)
+	
+	mainEnv = MainEnv(duration, interval)
+	mainEnv.start(input)
+
+
```
Comment (adbh94) [inline simulator.py:54]: Nit: extra newline.

**Comment by adbh94** [comment]
Looks like you are using tabs rather than 4-space indents in simulator.py. After fixing this and my other two comments, environment lgtm.

**Comment by sdhpzhtk** [comment]
Done. 

郑韬堃 Taokun (Xander) Zheng

Email: tzheng@caltech.edu
Mobile: 1-626-354-2788

Address:
California Institute of Technology
MSC 1076, Pasadena, CA 91126

**Comment by aman-agarwal** [comment]
@sdhpzhtk This branch needs to be merged from the command line.

**Pull request merged**